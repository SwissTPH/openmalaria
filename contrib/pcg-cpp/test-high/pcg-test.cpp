/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
