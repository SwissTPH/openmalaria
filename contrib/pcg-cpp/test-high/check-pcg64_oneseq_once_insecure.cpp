#define RNG pcg64_oneseq_once_insecure
#define TWO_ARG_INIT 0

#include "pcg-test.cpp"

