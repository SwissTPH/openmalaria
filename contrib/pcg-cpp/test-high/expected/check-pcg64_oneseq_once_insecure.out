pcg64_oneseq_once_insecure:
      -  result:      64-bit unsigned int
      -  period:      2^64
      -  size:        8 bytes

Round 1:
  64bit: 0x27a53829edf003a9 0xdf28458e5c04c31c 0x2756dc550bc36037
	 0xa10325553eb09ee9 0x40a0fccb8d9df09f 0x5c2047cfefb5e9ca
  Again: 0x27a53829edf003a9 0xdf28458e5c04c31c 0x2756dc550bc36037
	 0xa10325553eb09ee9 0x40a0fccb8d9df09f 0x5c2047cfefb5e9ca
  Coins: TTHHHTHHTTHHHTHTTHHTHHTTHHHTTHTTTTHHHTTHTHHHHTHTTTTTTHHHTTTHTTTTT
  Rolls: 6 4 3 6 3 4 6 6 2 3 5 6 1 4 3 6 4 1 5 2 3 4 2 5 4 5 3 6 6 4 4 2 2
   -->   rolling dice used 33 random numbers
  Cards: 3d Ah 7d Kh Td Th 8h 5d Ac 7c Ts As 7s 2s 2c 4h 8s Kc 3c 5h 6s 9c
	 4s Js Kd Qc 3h 2d Jh 6d 7h 3s 9d Tc 5s 9s 9h 4d Ks Jd Qd Ad Qs Jc
	 8d 4c 6c Qh 8c 5c 6h 2h

Round 2:
  64bit: 0xbf1a2718f46a9ab3 0x400c010fa404b25f 0x3e6dcc00e4e9238d
	 0x2ef92088e248ee1f 0xc6ba2d56b70972a8 0x7e11c48b5ee4b511
  Again: 0xbf1a2718f46a9ab3 0x400c010fa404b25f 0x3e6dcc00e4e9238d
	 0x2ef92088e248ee1f 0xc6ba2d56b70972a8 0x7e11c48b5ee4b511
  Coins: HHTHTHTTHTTHTTTTHTTHHHTHHTHTHHHHTHHTTTTTHTHTTHHTTTHTTHTHTTHTTHTHT
  Rolls: 5 1 2 3 6 1 5 2 3 2 2 6 4 6 5 6 1 5 4 3 1 1 5 3 4 5 5 2 3 4 2 3 5
   -->   rolling dice used 33 random numbers
  Cards: 9d Qs 4c 3s 7h 8d 3h Qd Jd 6s Ad Jh 5d Js 5s 8c 4h 5c 7s 8h Kc 9h
	 Ac 9c 3c 2s Tc 7d Ah 6d Jc Ks 2d 2c 9s Th 8s Kh Ts 4d 6h Td Qh Kd
	 3d 6c 2h 5h 7c As Qc 4s

Round 3:
  64bit: 0xa23134099c2db545 0xfd4094498a894a69 0x1199f424f91baa00
	 0x6c1c7a2896530d0b 0x1b9ca95e37253136 0xc05c20630e73d51f
  Again: 0xa23134099c2db545 0xfd4094498a894a69 0x1199f424f91baa00
	 0x6c1c7a2896530d0b 0x1b9ca95e37253136 0xc05c20630e73d51f
  Coins: HHTTHTTTTTTHHHTHHHTHHTTTHTTHHHHHTHTTHHHTHTTTHHHHTTTTHHTTHTTHHHHTT
  Rolls: 5 3 5 1 1 5 6 1 6 4 5 4 5 3 6 1 6 4 6 3 2 4 3 3 2 5 4 4 6 3 6 4 1
   -->   rolling dice used 33 random numbers
  Cards: Ac 7h Ts Kc 9d 7s Js 2h Th 9s 2s 6h 3h Jc 3s 2c 5h 5d Qd 8h 3d 5s
	 6c Qh 7c 8d 7d Td As Ah Ad 6s Ks 4c Jd 9c 3c 4d Jh 9h Qs Qc 2d Kh
	 5c 8c 8s 6d 4h 4s Tc Kd

Round 4:
  64bit: 0x61748befe0c9e052 0x9261b659354e4790 0xf5760c6a80d81414
	 0xe9f8cd09b4a145b6 0x63cea77be983a1c2 0xff19b1c85857ee53
  Again: 0x61748befe0c9e052 0x9261b659354e4790 0xf5760c6a80d81414
	 0xe9f8cd09b4a145b6 0x63cea77be983a1c2 0xff19b1c85857ee53
  Coins: THTHHHTHHHTTHTHTTTHHTHTTHTTHHHTHTHTHHHTTTHHHHHHHHTHTHHHHHTTTTTTTT
  Rolls: 6 6 1 3 3 2 6 5 5 2 3 4 5 3 3 6 6 5 6 1 4 3 5 3 1 2 4 2 2 2 6 6 1
   -->   rolling dice used 33 random numbers
  Cards: Jd 6h 5d 8s 3c Qd 2c 9s Ks 4h Kd 3h 6s 5c Jh Tc 4c Kh 2s 7d 8d Qh
	 As 9c Th 6c 9h 2h Qs 8c 7h 4s Js Ad 3d Jc Ac 7c 5h 2d Qc 6d Ah 7s
	 Ts 3s 5s 8h Kc Td 9d 4d

Round 5:
  64bit: 0x36288c2f6a1fcaba 0x9761b19d77017828 0x4192af9e0d374ee5
	 0x3d3134763fbd6bbf 0x1fb46e26bd261bd6 0x2c22c75b821893ac
  Again: 0x36288c2f6a1fcaba 0x9761b19d77017828 0x4192af9e0d374ee5
	 0x3d3134763fbd6bbf 0x1fb46e26bd261bd6 0x2c22c75b821893ac
  Coins: HHHHHHTTHTHHTTHHHTTTHHHTHHTHHHHHHTTTHTTTHTTTTTTHTTTHTTHTTHTTHHHHH
  Rolls: 6 4 3 4 5 5 2 3 1 6 6 6 6 4 4 4 1 2 2 5 4 1 5 2 3 4 1 4 1 5 1 3 5
   -->   rolling dice used 33 random numbers
  Cards: 9s Th 9d 8c Jc Td 7c 3c 7h 3h 7d 3s 8h Qs 6c Qd Ah Ad 4d 5h Jd Kh
	 8d Qc Ts 6s 9c 5s As Ac 4s 6h 3d Ks 4h 5d 2h 5c 2s 2c Tc Jh 9h Js
	 4c 6d 7s 2d Kc 8s Kd Qh

